command/auto_complete.go

```diff
@@ -0,0 +1,247 @@
+	isOnlyRemote = isOnlyRemote || isOnlyBucket
+	return func(ctx *cli.Context) {
+		var arg string
+		args := ctx.Args()
+		l := args.Len()
+
+		if l > 0 {
+			arg = args.Get(l - 1)
+		}
+
+		// argument may start with a quotation mark, in this case we want to trim that before
+		// checking if it has prefix s3://
+		// Beware that we only want to trim the first char, not all of the leading
+		// quotation marks, because those quotation marks may be actual charactes.
+		if strings.HasPrefix(arg, "'") {
+			arg = strings.TrimPrefix(arg, "'")
+		} else {
+			arg = strings.TrimPrefix(arg, "\"")
+		}
+
+		if isOnlyRemote || strings.HasPrefix(arg, "s3://") {
+			if strings.HasPrefix(arg, "-") {
+				cli.DefaultCompleteWithFlags(cmd)(ctx)
+				return
+			}
+			printS3Suggestions(ctx, arg, isOnlyBucket)
+			return
+		}
+
+		cli.DefaultCompleteWithFlags(cmd)(ctx)
+	}
+}
+
+// it returns a complete function which prints the argument, itself, which is to be completed.
+// If the argument is empty string it uses the defaultCompletions to make suggestions.
+func ineffectiveCompleteFnWithDefault(cmd *cli.Command, defaultCompletions ...string) func(ctx *cli.Context) {
+	return func(ctx *cli.Context) {
+		var arg string
+		args := ctx.Args()
+		if args.Len() > 0 {
+			arg = args.Get(args.Len() - 1)
+		}
+		if arg == "" {
+			for _, str := range defaultCompletions {
+				fmt.Println(formatSuggestionForShell(str, str))
+			}
+		} else if strings.HasPrefix(arg, "-") {
+			cli.DefaultCompleteWithFlags(cmd)(ctx)
+		} else {
+			fmt.Println(formatSuggestionForShell(arg, arg))
+		}
+	}
+}
+
+func printS3Suggestions(ctx *cli.Context, arg string, isOnlyBucket bool) {
+	c := ctx.Context
+	u, err := url.New(arg)
+	if err != nil {
+		u = &url.URL{Type: 0, Scheme: "s3"}
+	}
+
+	client, err := storage.NewRemoteClient(c, u, NewStorageOpts(ctx))
+	if err != nil {
+		return
+	}
+
+	if u.Bucket == "" || (u.IsBucket() && !strings.HasSuffix(arg, "/")) || isOnlyBucket {
+		printListBuckets(c, client, u, arg)
+	} else {
+		printListNURLSuggestions(c, client, u, 20, arg)
+	}
+}
+
+func printListBuckets(ctx context.Context, client *storage.S3, u *url.URL, argToBeCompleted string) {
+	buckets, err := client.ListBuckets(ctx, u.Bucket)
+	if err != nil {
+		return
+	}
+
+	for _, bucket := range buckets {
+		fmt.Println(formatSuggestionForShell("s3://"+bucket.Name+"/", argToBeCompleted))
+	}
+}
+
+func printListNURLSuggestions(ctx context.Context, client *storage.S3, u *url.URL, count int, argToBeCompleted string) {
+	if u.IsBucket() {
+		var err error
+		u, err = url.New(u.Absolute() + "/")
+		if err != nil {
+			return
+		}
+	}
+
+	i := 0
+	for obj := range (*client).List(ctx, u, false) {
+		if i > count {
+			break
+		}
+		if obj.Err != nil {
+			return
+		}
+		fmt.Println(formatSuggestionForShell(obj.URL.Absolute(), argToBeCompleted))
+		i++
+	}
+}
+
+func installCompletionHelp(shell string) {
```

Comment:
It does not really install right? 😆

Reply:
No it doesn't, it only prints the corresponding instructions and shell script 😅
Since the flag name was `install-completion` I named the function as `installCompletionHelp`, not a good naming.
I  renamed it to `printAutocompletionInstructions` now.